torch/_inductor/fx_passes/split_cat.py

```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
             )
         split_node.replace_all_uses_with(new_split_node)
         new_split_node.meta.update(split_node.meta)
         graph.erase_node(split_node)
         counters["inductor"]["split_cat_norm"] += 1
 
 
+class TorchSplit(CallFunction):
+    """
+    Matches a call to torch.split if it is in a normalized form. Ensures that all users of
+    splits are unique getitems.
+    """
+
+    def __init__(self, arg):
+        # using KeywordArg("dim") for `dim` checks they all match
+        super().__init__(
+            torch.split, arg, Arg(), _users=MULTIPLE, dim=KeywordArg("dim")
+        )
+
+    def _match(self, node: torch.fx.Node, ctx: MatchContext):
+        m = super()._match(node, ctx)
+        if not m:
+            return m
+        split_sections = get_arg_value(node, 1, "split_size_or_sections")
+        if not isinstance(split_sections, (list, tuple)):
+            return FailedMatch("split not normalized")
+        # check users are all unique getitems
+        seen_idxs = set()
+        for user in node.users:
+            if not match_node(user, {operator.getitem}):
+                # This should ideally never happen. Split user should always be a getitem
+                return FailedMatch(f"user of split not a getitem: {user}")
+            if user.args[1] in seen_idxs:
+                return FailedMatch(f"duplicate getitem {user.args[1]}")
+            if user.args[-1] < 0:
+                # This shouldn't ideally happen as dynamo normalizes indexes to positive
+                return FailedMatch("negative index")
+            seen_idxs.add(user.args[1])
+        return m
+
+
+class MergeConsecutiveSplits(PatternEntry):
+    def apply(self, match: Match, graph: torch.fx.Graph, node: torch.fx.Node):
+        first_split = node.args[0].args[0]
+        first_split_sections = get_arg_value(first_split, 1, "split_size_or_sections")
+        next_split_sections = get_arg_value(node, 1, "split_size_or_sections")
+        next_split_index = node.args[0].args[1]
+
+        new_split_sections = list(first_split_sections)
+        new_split_sections[
+            next_split_index : next_split_index + 1
+        ] = next_split_sections
+
+        first_split_input = first_split.args[0]
+        first_split_dim = get_arg_value(first_split, 2, "dim")
+
+        to_remove = []
+
+        with graph.inserting_after(first_split_input):
+            # Add the new split node
+            new_split = graph.call_function(
+                torch.split,
+                args=(first_split_input, new_split_sections),
+                kwargs={"dim": first_split_dim},
+            )
+            first_split_num_to_user = {
+                get_arg_value(user, 1): user for user in first_split.users.keys()
+            }
+
+            new_split_num = 0
+            for split_num in range(len(first_split_sections)):
+                if split_num not in first_split_num_to_user:
+                    new_split_num += 1
```

Comment:
Is this assuming the getitems are all in order and not having any gaps?

Reply:
I am iterating over `first_split_sections` which is the sections of the split, not getitems. The condition `if split_num not in first_split_num_to_user` should only hit when there is a gap in getitems. But, let me add test cases explicitly verifying the gap and out-of-order cases